cpp/include/cudf/table/experimental/row_operators.cuh

```diff
@@ -52,12 +54,84 @@
 #include <memory>
 #include <optional>
 #include <type_traits>
+#include <unordered_set>
 #include <utility>
+#include <variant>
 
 namespace CUDF_EXPORT cudf {
 
 namespace experimental {
 
+/// Type identity type from C++20
+template <typename T>
+using type_identity_t = T;
+
+/**
+ * @brief Recursive template to apply type transformations sequentially. Transformations are
+ * applied first-to-last in the order specified.
+ *
+ * @tparam T Base type or initial type on which transformations are applied
+ * @tparam Rest Type transformations to apply
+ */
+template <typename T, template <typename> typename... Rest>
+struct transform_sequence;
+
+/// @copydoc transform_sequence
+template <typename T, template <typename> typename First, template <typename> typename... Rest>
+struct transform_sequence<T, First, Rest...> {
+  using type =
+    typename transform_sequence<First<T>, Rest...>::type;  ///< Resolved type after transformations
+};
+
+/// @copydoc transform_sequence
+template <typename T>
+struct transform_sequence<T> {
+  using type = T;  ///< The underlying type
+};
+
+/**
+ * @brief Helper alias for transform_sequence
+ */
+template <typename T, template <typename> typename... Rest>
+using transform_sequence_t = typename transform_sequence<T, Rest...>::type;
+
+/**
+ * @brief Void dispatcher helper
+ */
+template <bool B, typename T>
+using dispatch_void_conditional_t = std::conditional_t<B, void, T>;
+
+/**
+ * @brief Void dispatcher generator
+ */
+template <typename... Types>
+struct dispatch_void_conditional_generator {
+  /// The underlying type
+  template <typename T>
+  using type = dispatch_void_conditional_t<std::disjunction<std::is_same<T, Types>...>::value, T>;
+};
+
+/**
+ * @brief Returns `void` if it's a nested type
+ */
+template <typename T>
+using dispatch_void_if_nested_t =
+  dispatch_void_conditional_generator<id_to_type<type_id::STRUCT>,
+                                      id_to_type<type_id::LIST>>::type<T>;
+
+/**
+ * @brief Returns `void` if it's a compound type
+ *
+ */
+template <typename T>
+using dispatch_void_if_compound_t =
+  dispatch_void_conditional_generator<id_to_type<type_id::STRUCT>,
+                                      id_to_type<type_id::LIST>,
+                                      id_to_type<type_id::DECIMAL128>,
+                                      id_to_type<type_id::DECIMAL64>,
+                                      id_to_type<type_id::DECIMAL32>,
```

Comment:
decimals are fixed-width types instead of compound. If this is intended, we probably want to use another term here.

Reply:
Out of curiosity, Will removing the decimal from the current compound definition impact performance?

Reply:
I believe row-operators may benefit being dispatched using the `dispatch_storage_type` defined here: https://github.com/rapidsai/cudf/blob/44a9c10105ab06538264e727188a04d623b0811e/cpp/include/cudf/utilities/type_dispatcher.hpp#L212 
Since we do not compare columns of different types, the scale values are the same and so the compare on storage type for decimals/fixed-point will have the same result with less generated code.
Here are a few examples of dispatching with the `dispatch_storage_type`:
https://github.com/rapidsai/cudf/blob/44a9c10105ab06538264e727188a04d623b0811e/cpp/src/sort/sort_column.cu#L46
https://github.com/rapidsai/cudf/blob/44a9c10105ab06538264e727188a04d623b0811e/cpp/include/cudf/detail/gather.cuh#L664

Maybe this can be combined with the dispatch mechanism here and eliminate any need for special handling for fixed-point types.

Reply:
I measured this by selectively commenting out the types in `type_dispatcher.hpp` and I see non-trivial speedup difference. In both the measured benchmarks, other complex types (list, struct, string, dict) are commented out.
I think for future work it would be good to see if we are still latency bound.
[Decimal effect on Hash join occupancy - Sheet1.csv](https://github.com/user-attachments/files/17044427/Decimal.effect.on.Hash.join.occupancy.-.Sheet1.csv)

(Speedup numbers in sheet because I learnt later that nvbench compare script only supports json format)